File: qviability/core/autothreshold.py
"""Automatic threshold seeds.

The default method is MaxEntropy (Kapur), reproducing ImageJ's
``setAutoThreshold("MaxEntropy dark")`` so each image opens at the threshold
the user is already used to. Otsu / Yen / Li / IsoData are offered for
comparison via scikit-image.
"""

from __future__ import annotations

import numpy as np
from skimage.filters import (threshold_isodata, threshold_li, threshold_otsu,
                             threshold_yen)

# Ordered for the UI dropdown; MaxEntropy first = default.
METHOD_NAMES = ["MaxEntropy", "Otsu", "Yen", "Li", "IsoData"]


def _kapur_threshold(hist: np.ndarray) -> int:
    """Kapur maximum-entropy threshold over a 256-bin histogram.

    Returns the bin index that maximizes the sum of background and foreground
    Shannon entropies — the same criterion as ImageJ's MaxEntropy.
    """
    hist = hist.astype(np.float64)
    total = hist.sum()
    if total <= 0:
        return 0
    p = hist / total

    cum_p = np.cumsum(p)                      # P1(t): background mass
    with np.errstate(divide="ignore", invalid="ignore"):
        plogp = np.where(p > 0, p * np.log(p), 0.0)
    cum_plogp = np.cumsum(plogp)
    total_plogp = cum_plogp[-1]

    best_t, best_val = 0, -np.inf
    for t in range(256):
        p1 = cum_p[t]
        p2 = 1.0 - p1
        if p1 < 1e-12 or p2 < 1e-12:
            continue
        h_back = np.log(p1) - cum_plogp[t] / p1
        h_fore = np.log(p2) - (total_plogp - cum_plogp[t]) / p2
        val = h_back + h_fore
        if val > best_val:
            best_val, best_t = val, t
    return best_t


def threshold_maxentropy(arr: np.ndarray, maxval: int) -> int:
    """MaxEntropy threshold for *arr*, returned as a raw pixel value in [0, maxval]."""
    hist, _ = np.histogram(arr, bins=256, range=(0, maxval + 1))
    t_bin = _kapur_threshold(hist)
    # Kapur splits at t_bin; the bright foreground starts just above it.
    raw = (t_bin + 1) / 256.0 * (maxval + 1)
    return int(min(round(raw), maxval))


def auto_threshold(arr: np.ndarray, maxval: int, method: str = "MaxEntropy") -> int:
    """Compute an auto-threshold for *arr* using *method*.

    Always returns a sensible int in [0, maxval]; falls back to the mean if a
    method fails on a degenerate (e.g. flat) image.
    """
    try:
        if method == "MaxEntropy":
            return threshold_maxentropy(arr, maxval)
        if method == "Otsu":
            value = threshold_otsu(arr)
        elif method == "Yen":
            value = threshold_yen(arr)
        elif method == "Li":
            value = threshold_li(arr)
        elif method == "IsoData":
            value = threshold_isodata(arr)
        else:
            return threshold_maxentropy(arr, maxval)
        return int(min(max(round(float(value)), 0), maxval))
    except Exception:
        return int(min(max(round(float(arr.mean())), 0), maxval))

File: web/src/core/autothreshold.ts
// Port of qviability/core/autothreshold.py.
//
// MaxEntropy (Kapur) is the default and is ported here exactly. Otsu is a
// first-cut over the 0..maxval histogram. Yen / Li / IsoData (currently from
// scikit-image) and exact-skimage Otsu parity are Phase 3 — until then the
// dispatcher falls back to the mean for those, matching the Python except path.

import { histogram256 } from "./imageIo";

export const METHOD_NAMES = ["MaxEntropy", "Otsu", "Yen", "Li", "IsoData"];

/** Python's round() — round half to even. */
function pyRound(x: number): number {
  const fl = Math.floor(x);
  const diff = x - fl;
  if (diff < 0.5) return fl;
  if (diff > 0.5) return fl + 1;
  return fl % 2 === 0 ? fl : fl + 1;
}

/**
 * Kapur maximum-entropy threshold over a 256-bin histogram. Returns the bin
 * index maximizing the sum of background and foreground Shannon entropies.
 */
export function kapurThreshold(hist: Float64Array): number {
  let total = 0;
  for (let i = 0; i < 256; i++) total += hist[i];
  if (total <= 0) return 0;

  const p = new Float64Array(256);
  for (let i = 0; i < 256; i++) p[i] = hist[i] / total;

  const cumP = new Float64Array(256);
  const cumPlogp = new Float64Array(256);
  let accP = 0;
  let accPlogp = 0;
  for (let i = 0; i < 256; i++) {
    accP += p[i];
    cumP[i] = accP;
    accPlogp += p[i] > 0 ? p[i] * Math.log(p[i]) : 0;
    cumPlogp[i] = accPlogp;
  }
  const totalPlogp = cumPlogp[255];

  let bestT = 0;
  let bestVal = -Infinity;
  for (let t = 0; t < 256; t++) {
    const p1 = cumP[t];
    const p2 = 1.0 - p1;
    if (p1 < 1e-12 || p2 < 1e-12) continue;
    const hBack = Math.log(p1) - cumPlogp[t] / p1;
    const hFore = Math.log(p2) - (totalPlogp - cumPlogp[t]) / p2;
    const val = hBack + hFore;
    if (val > bestVal) {
      bestVal = val;
      bestT = t;
    }
  }
  return bestT;
}

/** MaxEntropy threshold for `data`, as a raw pixel value in [0, maxval]. */
export function thresholdMaxEntropy(
  data: Uint8Array | Uint16Array,
  maxval: number,
): number {
  const hist = histogram256(data, maxval);
  const tBin = kapurThreshold(hist);
  const raw = ((tBin + 1) / 256.0) * (maxval + 1);
  return Math.min(pyRound(raw), maxval);
}

/** First-cut Otsu over the 0..maxval 256-bin histogram (exact parity: Phase 3). */
export function otsu(data: Uint8Array | Uint16Array, maxval: number): number {
  const hist = histogram256(data, maxval);
  let total = 0;
  let sum = 0;
  for (let i = 0; i < 256; i++) {
    total += hist[i];
    sum += i * hist[i];
  }
  if (total === 0) return 0;

  let sumB = 0;
  let wB = 0;
  let bestVar = -1;
  let bestBin = 0;
  for (let t = 0; t < 256; t++) {
    wB += hist[t];
    if (wB === 0) continue;
    const wF = total - wB;
    if (wF === 0) break;
    sumB += t * hist[t];
    const mB = sumB / wB;
    const mF = (sum - sumB) / wF;
    const between = wB * wF * (mB - mF) * (mB - mF);
    if (between > bestVar) {
      bestVar = between;
      bestBin = t;
    }
  }
  // Map bin back to a raw value in [0, maxval].
  const raw = (bestBin / 255.0) * maxval;
  return Math.min(pyRound(raw), maxval);
}

/**
 * Compute an auto-threshold for `data` using `method`. Always returns a
 * sensible int in [0, maxval]; falls back to the mean for not-yet-ported
 * methods or on failure.
 */
export function autoThreshold(
  data: Uint8Array | Uint16Array,
  maxval: number,
  method = "MaxEntropy",
): number {
  try {
    if (method === "MaxEntropy") return thresholdMaxEntropy(data, maxval);
    if (method === "Otsu") return otsu(data, maxval);
    // Yen / Li / IsoData: Phase 3.
    throw new Error(`method not yet ported: ${method}`);
  } catch {
    let sum = 0;
    for (let i = 0; i < data.length; i++) sum += data[i];
    const mean = data.length ? sum / data.length : 0;
    return Math.min(Math.max(pyRound(mean), 0), maxval);
  }
}
